File: daiphat-fe/src/types/transaction.type.ts
export enum PaymentGateway {
    PAYOS = 'PAYOS'
}

export enum TransactionStatus {
    PENDING = 'PENDING',
    COMPLETED = 'COMPLETED',
    FAILED = 'FAILED',
    CANCELLED = 'CANCELLED',
    REFUNDED = 'REFUNDED'
}

export enum TransactionType {
    OFFLINE = 'OFFLINE',
    ONLINE = 'ONLINE',
    REFUND = 'REFUND'
}

export interface TransactionResponse {
    id: number;
    /** Present for payment transactions; null/omitted for refund payouts (linked via refundRequestId). */
    orderId?: string;
    refundRequestId?: number;
    amount: number;
    gateway: PaymentGateway;
    gatewayOrderCode: number;
    paymentRef: string;
    status: TransactionStatus;
    type: TransactionType;
    paidAt?: string;
    cancelledAt?: string;
    failureReason?: string;
    paymentEvidenceUrl?: string;
    paymentBy?: string;
    note?: string;
}

export interface ProcessPaymentRequest {
    transactionId: number;
    gateway: PaymentGateway;
}

export interface CancelPaymentRequest {
    transactionId: number;
    gateway: PaymentGateway;
    reason?: string;
}

export interface PaymentResult {
    transactionId: number;
    type?: string;
    gateway: PaymentGateway;
    gatewayOrderCode: number;
    paymentRef: string;
    checkoutUrl: string;
    status: string;
    qrCode?: string | null;
    accountNumber?: string | null;
    accountName?: string | null;
    amount?: number | null;
    description?: string | null;
    bin?: string | null;
    expiredAt?: number | null;
}

export interface PendingPaymentCountdownResult {
    orderId: string;
    remainingSeconds: number;
    expiresAt: string | null;
    expired: boolean;
}

export interface PendingPaymentReminderResponse {
    orderId: string;
    orderCode: string;
    totalAmount: number;
    remainingSeconds: number;
    expiresAt: string | null;
    expired: boolean;
    transactionId: number | null;
    gateway: PaymentGateway;
}

File: daiphat_mobile/lib/src/features/checkout/models/transaction_type.dart
enum PaymentGateway {
  payos('PAYOS');

  final String value;
  const PaymentGateway(this.value);

  static PaymentGateway fromValue(String value) {
    return PaymentGateway.values.firstWhere(
      (e) => e.value == value,
      orElse: () => PaymentGateway.payos,
    );
  }
}

enum TransactionStatus {
  pending('PENDING'),
  completed('COMPLETED'),
  failed('FAILED'),
  cancelled('CANCELLED'),
  refunded('REFUNDED');

  final String value;
  const TransactionStatus(this.value);

  static TransactionStatus fromValue(String value) {
    return TransactionStatus.values.firstWhere(
      (e) => e.value == value,
      orElse: () => TransactionStatus.pending,
    );
  }
}

enum TransactionTypeEnum {
  offline('OFFLINE', 'Tiền mặt'),
  online('ONLINE', 'Chuyển khoản');

  final String value;
  final String label;
  const TransactionTypeEnum(this.value, this.label);

  static TransactionTypeEnum fromValue(String value) {
    return TransactionTypeEnum.values.firstWhere(
      (e) => e.value == value,
      orElse: () => TransactionTypeEnum.online,
    );
  }
}

class ProcessPaymentRequest {
  final int transactionId;
  final String gateway;

  const ProcessPaymentRequest({
    required this.transactionId,
    required this.gateway,
  });

  Map<String, dynamic> toJson() => {
    'transactionId': transactionId,
    'gateway': gateway,
  };
}

class PaymentResult {
  final int transactionId;
  final String gateway;
  final int? gatewayOrderCode;
  final String? paymentRef;
  final String? checkoutUrl;
  final String status;

  const PaymentResult({
    required this.transactionId,
    required this.gateway,
    this.gatewayOrderCode,
    this.paymentRef,
    this.checkoutUrl,
    required this.status,
  });

  factory PaymentResult.fromJson(Map<String, dynamic> json) {
    return PaymentResult(
      transactionId: json['transactionId'] as int? ?? 0,
      gateway: json['gateway']?.toString() ?? '',
      gatewayOrderCode: json['gatewayOrderCode'] as int?,
      paymentRef: json['paymentRef']?.toString(),
      checkoutUrl: json['checkoutUrl']?.toString(),
      status: json['status']?.toString() ?? '',
    );
  }
}

class EnumOption {
  final String value;
  final String label;

  const EnumOption({required this.value, required this.label});

  factory EnumOption.fromJson(Map<String, dynamic> json) {
    return EnumOption(
      value: json['value']?.toString() ?? '',
      label: json['label']?.toString() ?? '',
    );
  }
}
